[Mobile Payments] Fix IPP feedback survey banner display logic

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #8992
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR fixes various issues in the logic that determines if, and which, In Person Payments feedback survey banner to show the user. Full context: pdfdoF-2pO-p2

To fix these issues and show the correct banners, we've decided to patch up the iOS implementation to make it work as intended by:

1. Remove COD requirement for surveys 2 and 3.
2. Use Orders filtered to IPP-only for surveys 2 and 3 (instead of all WCPay Orders.)
3. Replace the filter which uses a server-localized string, `paymentMethodTitle` to use `paymentMethodID` instead, to avoid potential clashes with other plugins. (N.B. `paymentMethodID` is the same for IPP and web, so this isn’t sufficient to identify IPP transactions alone, but we use it in combination with the presence of receiptUrl, which is only set for IPP transactions within WCPay)

This PR is a step-by-step refactor, starting by fixing the tests and then progressively addressing the issues. A commit-by-commit review shouldn't be required, but there's some extra context in the commit messages if you choose to view it that way.

## Known issues
1. Locally stored orders are used, which means that getting the correct response requires the relevant orders to be loaded. In practice, this will often mean that the relevant orders must be in the first page (25) of orders.
2. Not all data is fetched from the same place, or at the same time. We constantly try to display the correct banner, so that results, in some cases, in one banner briefly being shown before the correct banner is shown. In practice, this seems very rare and I don't have any repro steps.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

Testing fully requires several stores, but the steps are simple:

1. Open the Order list
2. Observe the correct banner is shown

### Scenarios worth testing:

**Using a US or CA based store**
No WCPay IPP orders (web orders are fine)
With COD enabled
Shows banner with text "Share your own experience or how you collect in-person payments."

Same store with COD turned off (you can toggle it in the Payments menu of the app)
Doesn't show a banner

Using a US or CA based store with 1-9 WCPay IPP payments in the last 30 days
With COD enabled or disabled
Shows banner with text "Rate your first in-person payment experience."

Using a US or CA based store with 10 or more WCPay IPP payments
With COD enabled or disabled
Shows banner with text "Tell us all about your experience with in-person payments."

Using a US or CA based store with 1-9 WCPay IPP payments, all more than 30 days old
With COD enabled or disabled
Doesn't show a banner

**Using another country based store**
Using one of the above banner-displaying stores, temporarily switched to a non-US in wc-admin
Doesn't show a banner

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

joshheald committed:
Remove noisy teardown code

Setting the `selectedSiteSettings` takes time, and we already replace them in `setup` anyway.
Other properties are also replaced in `setup`
Resetting the `storageManager` adds console noise (as the CD stack is destroyed) and the storageManager is replaced anyway.

joshheald committed:
8992 Fix Cash on Delivery banner tests

Previously, the main assertions of the tests were not called, because the setup of the test did not include the COD payment gateway, so there was no call to the completion handler.

This fixes the COD tests so they correctly set up the view model to return a value, by setting a COD payment gateway and injecting the CardPresentPaymentsConfiguration for an IPP supported country.

Tests were renamed for more specificity about the conditions under test.

Currently, test_feedbackBannerSurveySource_Cash_on_Delivery_store_in_an_IPP_country_with_only_WCPay_web_orders_is_shown_inPersonPaymentsCashOnDelivery_survey is failing because we have not yet fixed the logic to use only IPP transactions, rather than all WCPay transactions

joshheald committed:
8992 Always call onCompletion for feedbackBanner

By allowing the survey to fall through, we missed issues in the tests. This change ensures we always call completion, to help with fixing the survey logic.

This change causes the four “displayed banner” tests to fail:

- test_feedbackBannerSurveySource_when_there_are_no_wcpay_orders_then_assigns_inPersonPaymentsCashOnDelivery_survey
- test_feedbackBannerSurveySource_when_there_is_one_wcpay_order_then_assigns_inPersonPaymentsCashOnDelivery_survey
- test_feedbackBannerSurveySource_when_there_are_less_than_ten_wcpay_orders_then_assigns_inPersonPaymentsFirstTransaction_survey
- test_feedbackBannerSurveySource_when_there_more_than_ten_wcpay_orders_then_assigns_inPersonPaymentsPowerUsers_survey

joshheald committed:
8992 Arrange tests of firstTransaction/powerUser surveys

Previously we did not inject the CardPresentPaymentsConfiguration, so it defaulted to an unsupported country and these tests recieved no banner in the response.

Also, we did not pass in the storageManager, so none of the orders we created were available to the code under test.

These tests still fail at this commit, because the logic under test (incorrectly) requires them to have COD enabled.

joshheald committed:
8992 Only require COD for the first survey

The inPersonPaymentsCashOnDelivery survey is intended to survey merchants who sell with COD (i.e. may be interested in IPP) but with a non-WCPay solution. This is supposed to target people who may switch to IPP with WCPay.

The other two surveys should not be limited to COD merchants, because not all WCPay merchants will want to offer that as a payment option on their website, even if they do collect IPP in some circumstances.

This also adds tests for the negative of this scenario, and when the store is in an unsupported country.

joshheald committed:
8992 Fix arrangement of receipt_url in tests

Prior to this, the `customFields` metadata weren’t saved to the test DB, so the tests which relied on `receipt_url` weren’t setting it correctly and would always fail regardless of what we do in the implementation.

joshheald committed:
8992 Remove duplictated `hasTransactions` conditoin

The base `if` for the `feedbackBannerSource` now excludes any stores which have no WCPay transactions from any later clauses, so we don’t need to check this in case 2.

joshheald committed:
8992 Update tests to include datePaid for recency

The conditions for surveys 2 and 3 rely on `datePaid`, which we were not setting in the tests, which means that the power users test

test_feedbackBannerSurveySource_when_there_are_more_than_ten_wcpay_orders_then_assigns_inPersonPaymentsPowerUsers_survey

was not arranging the data correctly, as it had no recent transactions.

joshheald committed:
8992 Require receipt_url for all orders wrt IPP surveys

All the IPP feedback survey logic should be based on WCPay IPP transactions. Previously, the results controllers did not include the required filter for the `receipt_url` metadata, so we (mostly) considered web orders as if they were IPP orders.

The exception to that was a post-fetch filter on the recentTransactions results.

This commit moves the filter to the predicate, and uses the predicate for both the results controllers (all WCPay IPP transactions, and recent WCPay IPP transactions.)

We never consider web orders, and now we can’t miss the filter out of our logic by mistake.

joshheald committed:
8992 remove unnecessary survey completionHandler

There’s no async work in the determination of the survey source, so the completion handler was adding unnecessary boilerplate and complexity.

joshheald committed:
Removed outdated warning about VM recreation

I’ve confirmed that the viewModel is now recreated when we switch sites.

joshheald committed:
8992 Push the feedback banner to the viewcontroller

Previously we pushed the instruction to show a feedback banner to the view, and then the view _pulled_ the banner type it should show. This was a little confusing and while it worked, it’s error prone.

This change pushes the banner tyep along with the instruction to show the banner.

joshheald committed:
8992 Improve display of feedback banners

Feedback banners were not reliably shown. This… isn’t the prettiest, because we don’t have a WCPayIPP related action to use as the prompt to update the banner type, so it uses the order action’s completion handler as a bit of a side-effect.

It’s not too bad, because the order action is the one that fetches the relevant data, and it’s only temporary anyway.

joshheald committed:
8992 Prevent firstIPP banner from showing for old orders

joshheald committed:
8992 add IPP feedback banner logic fix to 12.6 release

joshheald committed:
Merge branch 'trunk' into issue/8992-fix-ipp-survey-logic

joshheald commented:
@toupper @iamgabrielma @vanessamaartinstt doesn't necessarily need all of you to review, but any eyes on are welcome.

I'll do some more testing of it tomorrow, but I _think_ it works correctly now.

toupper commented:
@joshheald Thanks a lot for fixing this, the code looks good to me. When testing the cases, I got these results:

Using a US or CA based store
No WCPay IPP orders (web orders are fine)
With COD enabled
Shows banner with text "Share your own experience or how you collect in-person payments." ✅ 
Same store with COD turned off (you can toggle it in the Payments menu of the app)
Doesn't show a banner ✅ 

Using a US or CA based store with 1-9 WCPay IPP payments in the last 30 days
With COD enabled or disabled
Shows banner with text "Rate your first in-person payment experience." ✅ 

Using a US or CA based store with 10 or more WCPay IPP payments
With COD enabled or disabled
Shows banner with text "Tell us all about your experience with in-person payments." ✅ 

Using a US or CA based store with 1-9 WCPay IPP payments, all more than 30 days old
With COD enabled or disabled
Doesn't show a banner ❓ Unfortunately I couldn't test this one

Using another country based store
Using one of the above banner-displaying stores, temporarily switched to a non-US in wc-admin
Doesn't show a banner ✅

joshheald committed:
8992 Show OrderCreation banner when no IPP survey

By allowing `none` for the IPP banner source, we accidentally prevented the OrderCreation banner from showing, as we could “show” an empty IPP banner. This didn’t appear on screen, but dropped out of the top banner presentation code before the order creation banner got a chance to show.

This change removes the optionality from the TopBanner enum, so we either have an IPP banner to show, or we don’t (and we move to Order Creation), no half-way house.

This also fixes an issue where an (incorrectly shown) COD banner was not always cleared when all the data was fetched.

joshheald committed:
8992 Remove unneccesary constant

joshheald committed:
8992 Test historic IPP poweruser banner case

joshheald committed:
8992 Fix for IPP banner display edge case

When an IPP banner had been previously displayed, then a change in the underlying data led to an error, orderCreation, or no banner showing, the banner would not show again even when the underlying data got back to normal.

This change preserves the block on tracking this as a new display (as it can happen several times in quick succession) while ensuring the banner is shown again.

joshheald commented:
> @joshheald Thanks a lot for fixing this, the code looks good to me. When testing the cases, I got these results:
> 
> Using a US or CA based store No WCPay IPP orders (web orders are fine) With COD enabled Shows banner with text "Share your own experience or how you collect in-person payments." ✅ Same store with COD turned off (you can toggle it in the Payments menu of the app) Doesn't show a banner ✅
> 
> Using a US or CA based store with 1-9 WCPay IPP payments in the last 30 days With COD enabled or disabled Shows banner with text "Rate your first in-person payment experience." ✅
> 
> Using a US or CA based store with 10 or more WCPay IPP payments With COD enabled or disabled Shows banner with text "Tell us all about your experience with in-person payments." ✅
> 
> Using a US or CA based store with 1-9 WCPay IPP payments, all more than 30 days old With COD enabled or disabled Doesn't show a banner ❓ Unfortunately I couldn't test this one
> 
> Using another country based store Using one of the above banner-displaying stores, temporarily switched to a non-US in wc-admin Doesn't show a banner ✅

Thanks for the review @toupper 

I've retested all of these too, including the no-banner case. I found a couple more small issues, fixed them, and tested again. It's not perfect, but it can't be given our use of local stored data, and we knew that going in. It will show the correct banner most of the time now, though.
